File: Kotlin/src/des/Fiestel.kt
package des

import des.Config.EP
import des.Config.P32
import des.Config.SBOXES
import kotlin.math.pow

object Fiestel {
    private fun expandPermutation(x: ULong): ULong {
        var result = 0UL
        EP.forEachIndexed { index, p ->
            val currentBit = (x shr (32 - p)) and 1UL
            val shiftedX = currentBit shl (47 - index)
            result += shiftedX
        }
        return result
    }

    private fun permutation32(x: ULong): ULong {
        var result = 0UL
        P32.forEachIndexed { index, p ->
            val currentBit = (x shr (32 - p)) and 1UL
            val shiftedX = currentBit shl (31 - index)
            result += shiftedX
        }
        return result
    }

    private fun getSValueAt(x: Int, sBox: Array<IntArray>): ULong {
        val row = ((x and 0b100000) shr 4) or (x and 0b000001)
        val col = (x shr 1) and 0b001111
        return sBox[row][col].toULong()
    }

    private fun applySBoxes(x: ULong): ULong {
        var result = 0UL
        val mask: ULong = (2.0.pow(6)).toULong() - 1UL
        for (i in 1..8) {
            val temp = (x shr ((8 - i) * 6)) and mask
            val value = getSValueAt(temp.toInt(), SBOXES[i - 1])
            result = (result shl 4) or value
        }
        return result
    }

    fun apply(x: ULong, key: ULong): ULong {
        val msb = x shr 32
        val lsb = x and 0xFF_FF_FF_FFUL
        val ep = expandPermutation(lsb)
        val temp1 = ep xor key
        val temp2 = applySBoxes(temp1)
        val p32 = permutation32(temp2)
        val newMSB = msb xor p32
        return (newMSB shl 32) or lsb
    }
}

File: C/des/fiestel.c
#include "fiestel.h"

static unsigned long expand_permutation(unsigned long x) {
    unsigned long result = 0UL;
    for (int i = 0; i < BITS_EP; i++) {
        unsigned long current_bit = (x >> (BITS_P - EP[i])) & 1UL;
        unsigned long shifted_bit = current_bit << (BITS_EP - (i + 1));
        result += shifted_bit;
    }
    return result;
}

static unsigned long permutation(unsigned long x) {
    unsigned long result = 0UL;
    for (int i = 0; i < BITS_P; i++) {
        unsigned long current_bit = (x >> (BITS_P - P[i])) & 1UL;
        unsigned long shifted_bit = current_bit << (BITS_P - (i + 1));
        result += shifted_bit;
    }
    return result;
}

static unsigned long get_s_value_at(unsigned long x, const int s_box[4][16]) {
    int row = (int) (((x & 0b100000) >> 4) | (x & 0b000001));
    int col = (int) ((x >> 1) & 0b001111);
    return (unsigned long) s_box[row][col];
}

static unsigned long apply_s_boxes(unsigned long x) {
    unsigned long result = 0UL;
    unsigned long mask = (1 << 6) - 1UL;
    for (int i = 1; i <= 8; i++) {
        unsigned long temp = (x >> ((8 - i) * 6)) & mask;
        unsigned long value = get_s_value_at(temp, S_BOXES[i - 1]);
        result = (result << 4) | value;
    }
    return result;
}

unsigned long apply_fiestel_des(unsigned long x, unsigned long key) {
    unsigned long msb = x >> 32;
    unsigned long lsb = x & 0xFFFFFFFFUL;
    unsigned long ep = expand_permutation(lsb);
    unsigned long temp1 = ep ^ key;
    unsigned long temp2 = apply_s_boxes(temp1);
    unsigned long p = permutation(temp2);
    unsigned long new_msb = msb ^ p;
    return (new_msb << 32) | lsb;
}